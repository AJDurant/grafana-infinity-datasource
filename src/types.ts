import { DataQuery } from '@grafana/data';

export type dataPoint = [number | null, number];
export type timeSeriesResult = {
    target: string;
    datapoints: dataPoint[];
};
export type GrafanaTableColumn = {
    text: string;
    type: string;
};
export type GrafanaTableRowItem = (string | number | object | null);
export type GrafanaTableRow = GrafanaTableRowItem[];
export type tableResult = {
    columns: GrafanaTableColumn[];
    rows: GrafanaTableRow;
};
export type queryResult = timeSeriesResult | tableResult;

export interface ScrapColumn {
    selector: string;
    text: string;
    type: "string" | "number" | "timestamp";
}
export interface InfinityQuery extends DataQuery {
    type: 'json' | 'html' | 'csv';
    source: 'url' | 'inline';
    url: string;
    data: string;
    root_selector: string;
    columns: ScrapColumn[];
    format: 'table' | 'timeseries';
}
export const SCRAP_QUERY_TYPES = [{
    label: 'JSON',
    value: 'json'
}, {
    label: 'HTML',
    value: 'html'
}, {
    label: 'CSV',
    value: 'csv'
}];
export const SCRAP_QUERY_RESULT_FORMATS = [{
    label: 'Table',
    value: 'table'
}, {
    label: 'Time Series',
    value: 'timeseries'
}];
export const SCRAP_QUERY_SOURCES = [{
    label: 'URL',
    value: 'url'
}, {
    label: 'Inline',
    value: 'inline'
}];
export const SCRAP_QUERY_RESULT_COLUMN_FORMATS = [{
    label: 'String',
    value: 'string'
}, {
    label: 'Number',
    value: 'number'
}, {
    label: 'Timestamp',
    value: 'timestamp'
}];
