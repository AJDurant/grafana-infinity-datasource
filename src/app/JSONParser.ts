import { forEach, get } from 'lodash';
import { InfinityQuery, ScrapColumn } from "./../types";

export class JSONParser {
	private rows: any[];
	private series: any[];
	constructor(JSONResponse: object, private target: InfinityQuery, endTime?: Date) {
		this.rows = [];
		this.series = [];
		if (typeof JSONResponse === 'string') {
			JSONResponse = JSON.parse(JSONResponse)
		}
		if (target.root_selector) {
			JSONResponse = get(JSONResponse, target.root_selector);
		}
		if (Array.isArray(JSONResponse)) {
			forEach(JSONResponse, r => {
				const row: any[] = [];
				target.columns.forEach((c: ScrapColumn) => {
					row.push(get(r, c.selector, ""));
				});
				this.rows.push(row);
			});
			const NumbersColumns = target.columns.filter(t => t.type === 'number');
			NumbersColumns.forEach((metricColumn: ScrapColumn) => {
				forEach(JSONResponse, r => {
					let seriesName = target.columns.filter(t => t.type === 'string').map(c => r[c.selector]).join(' ');
					if (NumbersColumns.length > 1) {
						seriesName += ` ${metricColumn.text}`
					}
					const timestamp = endTime ? endTime.getTime() : new Date().getTime();
					this.series.push({
						target: seriesName,
						datapoints: [[get(r, metricColumn.selector), timestamp]]
					})
				})
			})
		} else {
			const row: any[] = [];
			target.columns.forEach((c: ScrapColumn) => {
				row.push(get(JSONResponse, c.selector, ""));
			});
			this.rows.push(row);
		}
	}
	toTable() {
		return {
			rows: this.rows.filter(row => row.length > 0),
			columns: this.target.columns
		}
	}
	toTimeSeries() {
		return this.series
	}
}