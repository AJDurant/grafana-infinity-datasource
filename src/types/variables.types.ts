import { SelectableValue } from '@grafana/data';
import { InfinityQuery } from './query.types';

//#region Variable Query
export type VariableTokenLegacy = 'Collection' | 'CollectionLookup' | 'Random' | 'Join' | 'UnixTimeStamp';
export type VariableQueryType = 'legacy' | 'infinity' | 'random';
export const variableQueryTypes: Array<SelectableValue<VariableQueryType>> = [
  {
    label: 'Infinity',
    value: 'infinity',
  },
  {
    label: 'Legacy',
    value: 'legacy',
  },
  {
    label: 'Random String',
    value: 'random',
  },
];
export type VariableQueryBase<T extends VariableQueryType> = { queryType: T };
export type VariableQueryLegacy = { query: string; infinityQuery?: InfinityQuery } & VariableQueryBase<'legacy'>;
export type VariableQueryInfinity = { infinityQuery: InfinityQuery; query?: string } & VariableQueryBase<'infinity'>;
export type VariableQueryRandom = { values: string[] } & VariableQueryBase<'random'>;
export type VariableQuery = VariableQueryLegacy | VariableQueryInfinity | VariableQueryRandom;
//#endregion
