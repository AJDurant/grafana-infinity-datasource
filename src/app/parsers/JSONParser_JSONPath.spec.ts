import { JSONParser } from './JSONParser';
import { InfinityQuerySources, InfinityQueryFormat, InfinityQueryType, ScrapColumnFormat } from './../../types';

const JSONResults = new JSONParser(
  {
    customers: [
      {
        name: 'mary',
        age: 22,
        gender: 'female',
      },
      {
        name: 'joseph',
        age: 41,
        gender: 'male',
      },
    ],
    premium_customers: [
      {
        name: 'john doe',
        age: 21,
        gender: 'male',
      },
    ],
  },
  {
    refId: '',
    type: InfinityQueryType.JSON,
    source: InfinityQuerySources.Inline,
    data: '',
    format: InfinityQueryFormat.Table,
    url: '',
    url_options: {
      method: 'GET',
    },
    root_selector: '$.*',
    columns: [
      {
        text: 'Name',
        type: ScrapColumnFormat.String,
        selector: 'name',
      },
    ],
  }
);
describe('JSON - JSONPath', () => {
  it('With Columns', () => {
    expect(JSONResults.toTable().rows.length).toBe(3);
    expect(JSONResults.toTable().rows[0].length).toBe(1);
    expect(JSONResults.toTable().rows[1][0]).toBe('joseph');
    expect(JSONResults.toTable().columns.length).toBe(1);
  });
});

const JSONResults1 = new JSONParser(
  {
    customers: [
      {
        name: 'mary',
        age: 22,
        gender: 'female',
      },
      {
        name: 'joseph',
        age: 41,
        gender: 'male',
      },
    ],
    premium_customers: [
      {
        name: 'john doe',
        age: 21,
        gender: 'male',
      },
    ],
  },
  {
    refId: '',
    type: InfinityQueryType.JSON,
    source: InfinityQuerySources.Inline,
    data: '',
    format: InfinityQueryFormat.Table,
    url: '',
    url_options: {
      method: 'GET',
    },
    root_selector: '$.*',
    columns: [
      {
        text: 'Name',
        type: ScrapColumnFormat.String,
        selector: 'name',
      },
      {
        text: 'Age',
        type: ScrapColumnFormat.Number,
        selector: 'age',
      },
    ],
  }
);
describe('JSON - JSONPath', () => {
  it('With multi Columns', () => {
    expect(JSONResults1.toTable().rows.length).toBe(3);
    expect(JSONResults1.toTable().rows[0].length).toBe(2);
    expect(JSONResults1.toTable().rows[2][0]).toBe('john doe');
    expect(JSONResults1.toTable().columns.length).toBe(2);
  });
});

const JSONResults2 = new JSONParser(
  {
    customers: [
      {
        name: 'mary',
        age: 22,
        gender: 'female',
      },
      {
        name: 'joseph',
        age: 41,
        gender: 'male',
      },
    ],
    premium_customers: [
      {
        name: 'john doe',
        age: 21,
        gender: 'male',
      },
    ],
  },
  {
    refId: '',
    type: InfinityQueryType.JSON,
    source: InfinityQuerySources.Inline,
    data: '',
    format: InfinityQueryFormat.Table,
    url: '',
    url_options: {
      method: 'GET',
    },
    root_selector: '$.customers',
    columns: [
      {
        text: 'Name',
        type: ScrapColumnFormat.String,
        selector: 'name',
      },
      {
        text: 'Age',
        type: ScrapColumnFormat.Number,
        selector: 'age',
      },
    ],
  }
);
describe('JSON - JSONPath', () => {
  it('With multi Columns', () => {
    expect(JSONResults2.toTable().rows.length).toBe(2);
    expect(JSONResults2.toTable().rows[0].length).toBe(2);
    expect(JSONResults2.toTable().rows[0][0]).toBe('mary');
    expect(JSONResults2.toTable().columns.length).toBe(2);
  });
});

const JSONResults3 = new JSONParser(
  {
    $customers: [
      {
        name: 'mary',
        age: 22,
        gender: 'female',
      },
      {
        name: 'joseph',
        age: 41,
        gender: 'male',
      },
    ],
    $premium_customers: [
      {
        name: 'john doe',
        age: 21,
        gender: 'male',
      },
    ],
  },
  {
    refId: '',
    type: InfinityQueryType.JSON,
    source: InfinityQuerySources.Inline,
    data: '',
    format: InfinityQueryFormat.Table,
    url: '',
    url_options: {
      method: 'GET',
    },
    root_selector: '$customers',
    columns: [
      {
        text: 'Name',
        type: ScrapColumnFormat.String,
        selector: 'name',
      },
      {
        text: 'Age',
        type: ScrapColumnFormat.Number,
        selector: 'age',
      },
      {
        text: 'Gender',
        type: ScrapColumnFormat.String,
        selector: 'gender',
      },
    ],
  }
);
describe('JSON - JSONPath', () => {
  it('With multi Columns', () => {
    expect(JSONResults3.toTable().rows.length).toBe(2);
    expect(JSONResults3.toTable().rows[0].length).toBe(3);
    expect(JSONResults3.toTable().rows[0][0]).toBe('mary');
    expect(JSONResults3.toTable().columns.length).toBe(3);
  });
});
