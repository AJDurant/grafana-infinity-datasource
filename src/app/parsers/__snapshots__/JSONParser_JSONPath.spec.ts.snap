// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON - JSONPath With Columns 1`] = `
Object {
  "columns": Array [
    Object {
      "selector": "name",
      "text": "Name",
      "type": "string",
    },
  ],
  "name": "",
  "rows": Array [
    Array [
      "mary",
    ],
    Array [
      "joseph",
    ],
    Array [
      "john doe",
    ],
  ],
}
`;

exports[`JSON - JSONPath With multi Columns 1`] = `
Object {
  "columns": Array [
    Object {
      "selector": "name",
      "text": "Name",
      "type": "string",
    },
    Object {
      "selector": "age",
      "text": "Age",
      "type": "number",
    },
  ],
  "name": "",
  "rows": Array [
    Array [
      "mary",
      22,
    ],
    Array [
      "joseph",
      41,
    ],
    Array [
      "john doe",
      21,
    ],
  ],
}
`;

exports[`JSON - JSONPath With multi Columns 2`] = `
Object {
  "columns": Array [
    Object {
      "selector": "name",
      "text": "Name",
      "type": "string",
    },
    Object {
      "selector": "age",
      "text": "Age",
      "type": "number",
    },
  ],
  "name": "",
  "rows": Array [
    Array [
      "mary",
      22,
    ],
    Array [
      "joseph",
      41,
    ],
  ],
}
`;

exports[`JSON - JSONPath With multi Columns 3`] = `
Object {
  "columns": Array [
    Object {
      "selector": "name",
      "text": "Name",
      "type": "string",
    },
    Object {
      "selector": "age",
      "text": "Age",
      "type": "number",
    },
    Object {
      "selector": "gender",
      "text": "Gender",
      "type": "string",
    },
  ],
  "name": "",
  "rows": Array [
    Array [
      "mary",
      22,
      "female",
    ],
    Array [
      "joseph",
      41,
      "male",
    ],
  ],
}
`;
